### PR TITLE
Create monitor.md

### PR DESCRIPTION
**Pull request by StreamingDataApps**
process monitor tutorial re-write
https://github.com/swimos/swimos-docs/issues/63

### CODE REVIEW

File: src/_tutorials/monitor.md [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+layout: page
+---
+
+## Overview
+
+We’ll be looking at a process monitoring <a href="https://github.com/swimos/tutorial-monitor">example</a> that provides data for individual machines in a cluster, as well as aggregate metrics for the cluster. The solution will consist of two separate applications. The first is a client application that runs on each machine in the cluster to collect monitoring information. The second is a SwimOS server application that processes results for each machine via a `MachineAgent`, and then aggregates cluster-level metrics across all machines via a `ClusterAgent`.
+
+## Monitoring Client
+
+The Monitoring Client relies on the OSHI java library to extract system information from each host machine in the cluster. OSHI is a free JNA-based (native) Operating system and Hardware Information library for Java. Each client periodically reports its operational health to the server. Since the server runs on top of the SwimOS runtime, the monitoring client uses `swim.ClientRuntime` to send commands bearing monitoring data updates.
```
Comment (SirCipher) [inline src/_tutorials/monitor.md:12]: Could you add a link to the OSHI repository?

Comment (SirCipher) [inline src/_tutorials/monitor.md:12]: `swim.client.ClientRuntime`

File: src/_tutorials/monitor.md [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+  public void pulse() {
+    Value usage = getUsage(timestamp);
+    this.warpRef.command(hostUri, nodeUri, Uri.parse("addUsage"), usage);
+  }
+
+  private Value getMemoryUtilization() {
+    return Record.create(6)
+    .slot("total", systemInfo.getHardware().getMemory().getTotal())
+    .slot("available", systemInfo.getHardware().getMemory().getAvailable())
+;
```
Comment (SirCipher) [inline src/_tutorials/monitor.md:35]: Formatting

File: src/_tutorials/monitor.md [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+  }
+
+  private Value getHardwareUtilization() {
+    Value memoryInfo = getMemoryUtilization();
+    return Record.of().slot("memory", memoryInfo);
+  }
+
+  private Value getUsage(final long timestamp) {
```
Comment (SirCipher) [inline src/_tutorials/monitor.md:43]: This timestamp is never used both here and in the code repository

File: src/_tutorials/monitor.md [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+
+    return Record.of()
+            .slot("timestamp", System.currentTimeMillis())
+            .slot("hardware", hardwareInfo);
+  }
+```
+
+### SwimMonitorClient
+
+In <a href="https://github.com/swimos/tutorial-monitor/blob/main/server/src/main/java/swim/monitor/client/SwimMonitorClient.java">`SwimMonitorClient`</a>, we will instantiate the SwimOS client runtime, the OSHI `SystemInfo`, and the `ProcessMonitor`.
```
Comment (SirCipher) [inline src/_tutorials/monitor.md:54]: "... which then starts the process monitor to feed data to the server"

File: src/_tutorials/monitor.md [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+
+```java
+  public static void main(String[] args) {
+
```
Comment (SirCipher) [inline src/_tutorials/monitor.md:58]: Empty line

File: src/_tutorials/monitor.md [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+}
+```
+
+This is invoked when setting the value of the `MachineAgent`’s status lane.
+
+```java
+  @SwimLane("status")
+  ValueLane<Value> status = this.<Value>valueLane()
+          .willSet(StatusComputer::computeSeverityFromStatus)
+          .didSet((nv, ov) -> {
```
Comment (SirCipher) [inline src/_tutorials/monitor.md:111]: For the tutorial and code repository, these parameters should be 'newValue' and 'oldValue' so it's immediately obvious what they are

File: src/_tutorials/monitor.md [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+- process information for each process running on the machine (static and dynamic)
+
+
+When static information is set, the <a href="https://github.com/swimos/tutorial-monitor/blob/main/server/src/main/java/swim/monitor/agent/MachineAgent.java">`MachineAgent`</a> registers with the cluster using the following command:
+
+```java
+command(CLUSTER_URI_PATTERN.apply("default"), ADD_MACHINE_CLUSTER_LANE_URI, Uri.form().mold(nodeUri()).toValue());
+```
+
+By registering, the `ClusterAgent` will be able to observe when monitoring data changes for all hosts that register.
```
Comment (SirCipher) [inline src/_tutorials/monitor.md:91]: I think that this line could be a bit clearer

File: src/_tutorials/monitor.md [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+
+Let's look how memory usage information is computed at the machine level based on monitoring input.
```
Comment (SirCipher) [inline src/_tutorials/monitor.md:93]: I think that these next two blocks should be switched around so the guide follows the flow of the data. I.e, "When the "addUsage" lane receives a new value, it invokes `computeStatusFromUsage`"

File: src/_tutorials/monitor.md [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+  @SwimLane("usage")
+  ValueLane<Value> usage = this.<Value>valueLane()
+          .didSet((newValue, oldValue) -> {
+            this.status.set(StatusComputer.computeStatusFromUsage(this.status.get(), newValue));
+          });
+```
+
+### Cluster Web Agent
+
+The <a href="https://github.com/swimos/tutorial-monitor/blob/main/server/src/main/java/swim/monitor/agent/ClusterAgent.java">`ClusterAgent`</a> receives monitoring data for all machines that register to it via a corresponding `MachineAgent`. When a machine registers using `ClusterAgent::addMachine` command lane, the `ClusterAgent` downlinks to the `MachineAgent`’s status lane via its `machines` `JoinValueLane`:
```
Comment (SirCipher) [inline src/_tutorials/monitor.md:131]: I'd add a link to the downlinks tutorial here

File: src/_tutorials/monitor.md [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+    int machineCount = 0;
+    double clusterAvgMemoryUsage = 0.0;
+    double clusterAvgMemoryUsage = 0.0;
+
+    float totalMemoryUsage = 0.0f, maxMemoryUsage = 0.0f;
+    int memoryCount = 0;
+
+    final Set<Value> keys = this.machines.keySet();
+    for (Value key : keys) {
+      machineCount++;
```
Comment (SirCipher) [inline src/_tutorials/monitor.md:179]: This isn't needed as it's the length of the machines set

File: src/_tutorials/monitor.md [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+  public static void main(String[] args) {
+
+    final ClientRuntime swimClient = new ClientRuntime();
+    swimClient.start();
+    final SystemInfo systemInfo = new SystemInfo();
+    startProcessMonitor(swimClient, systemInfo);
+  }
+
+  private static void startProcessMonitor(final ClientRuntime swimClient, final SystemInfo systemInfo) {
+    final ProcessMonitor processMonitor = new ProcessMonitor(swimClient, HOST_URI, systemInfo);
```
Comment (SirCipher) [inline src/_tutorials/monitor.md:66]: In the code repository, the `HOST_URI` is read from the args or defaults to 9001. But the server doesn't do the same as it just reads the recon file

File: src/_tutorials/monitor.md [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+swim-cli sync -h warp://localhost:9001 -n /cluster/abc -l statusHistory
+```
+
+## Running the UI
+
+Now, under the `/ui` folder under project root, open `index.html` as a local file in your web browser to see results from monitoring your locate machine populate a chart.
+
+## Demo Application
+
+A more complete of this source code exists as the swim-monitor demo can be found <a href="https://github.com/nstreamio/swim-monitor">here</a>.
```
Comment (SirCipher) [inline src/_tutorials/monitor.md:283]: Should this not be https://github.com/swimos/tutorial-monitor ?

File: src/_tutorials/monitor.md [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+
+```
+swim-cli sync -h warp://localhost:9001 -n /cluster/abc -l machines
+swim-cli sync -h warp://localhost:9001 -n /cluster/abc -l status
+swim-cli sync -h warp://localhost:9001 -n /cluster/abc -l statusHistory
+```
+
+## Running the UI
+
+Now, under the `/ui` folder under project root, open `index.html` as a local file in your web browser to see results from monitoring your locate machine populate a chart.
```
Comment (SirCipher) [inline src/_tutorials/monitor.md:279]: The server could serve the UI using the `UIRouter` kernel and then the user could just open the corresponding link

**Review by SirCipher — CHANGES_REQUESTED** [review]

**Comment by StreamingDataApps** [comment]
I've addressed the previous review comments. This PR is a bit cluttered, maybe I should open a new one?

**Review by SirCipher — APPROVED** [review]

**Pull request merged**